### DISCUSSION TITLE
CLI refactor

### DISCUSSION THREAD

**Discussion by kanishkatn** [ChainSafe/gossamer#3171] [General]
Currently, we have the following commands in gossamer
```
export
   Exports the configuration values to TOML configuration file. It takes in the values from flags, another toml and the default values, and creates a new toml file.
init
   Initialises a node with toml, db and genesis file.
account
   key management
build-spec
   Generates genesis JSON data, and can convert to raw genesis data 
import-runtime
   Appends the given .wasm runtime binary to a genesis file
import-state
   Import state from a JSON file and set it as the chain head state
prune
   Prunes the state trie
```

I've removed `export` in the new CLI and handling it as part of `init`. 

Here are the commands in Substrate
```
Commands:
  build-spec
          Build a chain specification
  check-block
          Validate blocks
  export-blocks
          Export blocks
  export-state
          Export the state of a given block into a chain spec
  import-blocks
          Import blocks
  purge-chain
          Remove the whole chain
  revert
          Revert the chain to a previous state
  benchmark
          Sub-commands concerned with benchmarking. The pallet benchmarking moved to the `pallet` sub-command
  host-perf-check
          Runs performance checks such as PVF compilation in order to measure machine capabilities of running a validator
  try-runtime
          Try some command against runtime state. Note: `try-runtime` feature must be enabled
  key
          Key management CLI utilities
  chain-info
          Db meta columns information
```

Creating this for discussions around the existing commands and the changes we'd like to see in the new CLI.

**Comment by EclesioMeloJunior** [comment]
I don't see how useful the `init` command is. We can start/run a node without necessarily calling the `init` command, so I we should remove it from the command set IMO

**Reply by kanishkatn** [reply]
Yes, you're right! It does initialise the node during `run` if it hasn't been initialised before.

I've added another use case for `init`, wherein it generates the config file using the default-values and the cli flags. The config file comes with a doc which we could use to further modify the values before running the node. 

For a node operator, I feel it is easier to work with the config file than learning about all the different flags. So, I've kept it in.

**Reply by EclesioMeloJunior** [reply]
100% agree over the usage of a config file, I use it all the time. So I think your purpose is to make this init command more or less equal to the `npm init` which provides a `package.json` with some values there, right? If so, in my opinion `gossamer init` is quite vague and looks like we are trying to "initialize" a node, so I would like to suggest a command like `gossamer config init {flags}`

**Reply by kanishkatn** [reply]
hmm, the `init` already does the following
- Takes in command line params and the `chain`, and creates the `toml` config
- Copies over the `toml` to the base-path
- Does the DB setup in the base-path

So, in a way it is initialising the node to be run at a later stage and not just creating the skeleton config. 

`tendermint` also has the `init` command, wherein it creates the config, private-key file, and few other files that are needed to run the node. We're gonna have to move the `node-key` to a file eventually as well.

**Reply by EclesioMeloJunior** [reply]
Yeah, but even after adding the new use case, as you described, I keep not seeing the point of how useful `init` is (imo), since as a node operator (used to run substrate node) I don't need an `init` to setup my node before running it. However, I totally agree about the usefulness of a `config.toml` and I see that could be useful and a good experience a command that only generates a TOML file without any other setup:

```sh
gossamer config init // generates `config.toml` using default values
gossamer --config config.toml // starts the node using the config.toml
```

**Reply by kanishkatn** [reply]
Yeah, no. I understand your point. 

I'm just saying that it isn't gonna create just the `config.toml` but also the `genesis.json` and `node-key.json` with room for more later if needed. 

Since it isn't just the `config.toml` that is being initialised, it doesn't make sense to rename the command.
